compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -1036,27 +1053,38 @@ abstract class GenJSCode extends plugins.PluginComponent
 
     // Constructor of a non-native JS class ------------------------------
 
-    def genJSClassConstructor(classSym: Symbol,
-        constructorTrees: List[DefDef]): js.MethodDef = {
+    def genJSClassCapturesAndConstructor(classSym: Symbol,
+        constructorTrees: List[DefDef]): (Option[List[js.ParamDef]], js.MethodDef) = {
       implicit val pos = classSym.pos
 
       if (hasDefaultCtorArgsAndRawJSModule(classSym)) {
         reporter.error(pos,
             "Implementation restriction: constructors of " +
             "non-native JS classes cannot have default parameters " +
             "if their companion module is JS native.")
-        js.MethodDef(static = false, js.StringLiteral("constructor"), Nil,
-            jstpe.AnyType, Some(js.Skip()))(
+        val ctorDef = js.MethodDef(static = false,
+            js.StringLiteral("constructor"), Nil, jstpe.AnyType,
+            Some(js.Skip()))(
             OptimizerHints.empty, None)
+        (None, ctorDef)
       } else {
-        withNewLocalNameScope {
+        withNewLocalNameScopeUsingJSSuperClassParamName {
```

Comment:
The difference between the two is that:
* `JSSuperClassParamName` is a capture param, referencing the superclass value of the enclosing `ClassDef`. It is only available in *instance members* (not static members) of the `ClassDef`, and therefore cannot be used in regular `genMethodBody`'s, because the `DefDef`s in a non-native JS class are compiled as static methods taking `this` as a parameter. It is only used for the dispatchers in `GenJSExports`, which are always used for instance members (methods and properties) and never for static methods.
* `jsSuperClassValueForSuperCall` is a mechanism to ensure that the static methods generated for `DefDef`s somehow receive a reference to the superclass value. Since they cannot use `JSSuperClassParamName`, they need to receive it in some other way. We make sure they get it through `lambdalift` doing its job, by explicitly accessing that superclass value as a term inside `innerOrLocalJSSuperCall`.

We cannot use a scoped variable because, being a class-level capture, it must be exactly the same in the entire class. But we generate a different local-name scope for each dispatcher method/property (as we should), so we need an external way to enforce that all the local-name scopes agree on the name of that class-level capture; they do not have the authority to choose it.